buildSrc/src/main/groovy/org/elasticsearch/gradle/NoticeTask.groovy

```diff
@@ -49,25 +55,38 @@ public class NoticeTask extends DefaultTask {
         dependencies.add(project)
     }
 
+    /** Add notices from an extra directory. */
+    public void extraLicensesDir(File extraLicensesDir) {
+        extraLicensesDirs.add(extraLicensesDir)
+    }
+
     @TaskAction
     public void generateNotice() {
         StringBuilder output = new StringBuilder()
         output.append(inputFile.getText('UTF-8'))
         output.append('\n\n')
-        Set<String> seen = new HashSet<>()
+        Map<String, File> seen = new TreeMap<>()
         for (Project dep : dependencies) {
-            File licensesDir = new File(dep.projectDir, 'licenses')
-            if (licensesDir.exists() == false) continue
-            licensesDir.eachFileMatch({ it ==~ /.*-NOTICE\.txt/ && seen.contains(it) == false}) { File file ->
-                String name = file.name.substring(0, file.name.length() - '-NOTICE.txt'.length())
-                appendFile(file, name, 'NOTICE', output)
-                appendFile(new File(file.parentFile, "${name}-LICENSE.txt"), name, 'LICENSE', output)
-                seen.add(file.name)
-            }
+            iterateLicensesDir(new File(dep.projectDir, 'licenses'), seen, true)
+        }
+        for (File extraLicensesDir : extraLicensesDirs) {
+            iterateLicensesDir(extraLicensesDir, seen, false)
+        }
+        for (File file : seen.values()) {
+            String name = file.name.substring(0, file.name.length() - '-NOTICE.txt'.length())
+            appendFile(file, name, 'NOTICE', output)
+            appendFile(new File(file.parentFile, "${name}-LICENSE.txt"), name, 'LICENSE', output)
         }
         outputFile.setText(output.toString(), 'UTF-8')
     }
 
+    static void iterateLicensesDir(File licensesDir, Map<String, File> seen, boolean isOptional) {
+        if (isOptional && licensesDir.exists() == false) return
```

Comment:
This task used to be lenient for the standard `${projectDir}/licenses` directories, so to remove the leniency here I had to add in extra checks in `PluginBuildPlugin.groovy` and `distribution/build.gradle`.